staging/src/k8s.io/apiserver/pkg/endpoints/filters/request_deadline.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"context"
+	"fmt"
+	"k8s.io/klog/v2"
+	"net/http"
+	"time"
+
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+// WithRequestDeadline determines the deadline of the given request and sets a new context
+// with the appropriate timeout.
+// timeout specifies the default request timeout value
+func WithRequestDeadline(handler http.Handler, longRunning request.LongRunningRequestCheck, timeout time.Duration) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		ctx := req.Context()
+
+		requestInfo, ok := request.RequestInfoFrom(ctx)
+		if !ok {
+			handleError(w, req, fmt.Errorf("no RequestInfo found in context, handler chain must be wrong"))
+			return
+		}
+		if longRunning(req, requestInfo) {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		if userProvidedTimeout, ok := parseTimeout(req); ok {
+			timeout = userProvidedTimeout
```

Comment:
we allow users to set longer timeouts? Did we used to allow them to set longer ones?

Reply:
@deads2k I did not find a code site where we check whether the user specified timeout exceeds the maximum allowed (`60s`), this is in keeping with the current behavior. I have made the following changes, let me know if this sounds good to you.
A: invalid value in `timeout` query parameter - we abort the request with `http.StatusBadRequest` 
B: user specified timeout exceeds the maximum request timeout `60s` -  we abort the request with `http.StatusBadRequest`

I should note that both are changes in behavior.

A:
before: default timeout kicked in and the request worked even though it has a bad user input.
now: the request will fail with a `http.StatusBadRequest` 
The API contract says timeout should have a valid specifier for time duration, so this change is restoring the behavior as per contract. but as a consequence we might see broken clients.

B: 
- before: the request timed out at `60s` with a timeout error. (no matter what the user specified timeout is, the timeout handler will timeout the request after `60s` anyway)
- now: the request will time out at `60s` as before but with a `http.StatusBadRequest` error.